lib/iris/experimental/ugrid.py

```diff
@@ -1680,102 +1679,316 @@ def remove(
-#         # choose the container wisely e.g., could be an dict..., also the self
-#         # attributes may need to be @property's that access the chosen _members container
-#
-#         # is this a list? as dict? a namedtuple? use case is self.add()
-#         self._members = []
-#
-#         if self.edge_node is not None:
-#             self._members.append(self.edge_node)
-#
-#     def __iter__(self):
-#         for member in self._members:
-#             yield member
-#
-#     def __getstate__(self):
-#         pass
-#
-#     def __setstate__(self, state):
-#         pass
-#
-#     def connectivity(self, **kwargs):
-#         # see Cube.coord for pattern, checking for a single result
-#         return self.connectivities(**kwargs)[0]
-#
-#     def connectivities(self, ...):
-#         # see Cube.coords for relevant patterns
-#         # return [ ... ]
-#         pass
-#
-#     def add(self, *args):
-#         # loop thru args and add (clobber)
-#         # adopt same philosophy as remove for adding connectivites with unsupported cf-role
-#         pass
-#
-#     def remove(self, ...):
-#         # needs to respect the minimum UGRID contract
-#         # use logging/warning to flag items not removed - highlight in doc-string
-#         # don't raise an exception
-#
-#     def __str__(self):
-#         pass
-#
-#     def __repr__(self):
-#         pass
-#
-#     def __eq__(self, other):
-#         # Full equality could be MASSIVE, so we want to avoid that.
-#         # Ideally we want a mesh signature from LFRic for comparison, although this would
-#         # limit Iris' relevance outside MO.
-#         # TL;DR: unknown quantity.
-#         raise NotImplemented
-#
-#     def __ne__(self, other):
-#         # See __eq__
-#         raise NotImplemented
-#
-#
-# class _Mesh2DConnectivityManager(_Mesh1DConnectivityManager):
-#     REQUIRED = (
-#         "face_node",
-#     )
-#     OPTIONAL = (
-#         "edge_node",
-#         "face_edge",
-#         "face_face",
-#         "edge_face",
-#         "boundary_node",
-#     )
-#     def __init__(self, face_node, edge_node=None, face_edge=None, face_face=None, edge_face=None, boundary_node=None):
-#         # required
-#         self.face_node = face_node
-#         self._members = [self.face_node]
-#
-#         # optionally required
-#         self.edge_node = edge_node
-#         # optional
-#         self.face_edge = face_edge
-#         self.face_face = face_face
-#         self.edge_face = edge_face
-#         self.boundary_node = boundary_node
-#
-#         # edge_node could be None here. are we okay with this pattern?
-#         super().__init__(edge_node)
-#
-#         # does order matter?
-#         self._members.extend([member for member in self.OPTIONAL if member is not None and member != "edge_node"])
+class _MeshConnectivityManagerMixin(ABC):
+    REQUIRED = ()
+    OPTIONAL = ()
+    NDIM = NotImplemented
+
+    @abstractmethod
```

Comment:
@trexfeathers There is no need to make the `__init__` abstract.

You're forcing the subclasses to implement it, and they are not specialising behaviour because you've decided to make things generic.

Reply:
Nice catch! A classic product of many local refactors. It's obvious now you've said it.

a8897abb48759742a48b84f8f82986c78bdb42b6